File: R/algorithmic_toolbox/25_arithmetic_expression.R
#!/usr/bin/env Rscript
Operations <- list("+"=.Primitive("+"), "-"=.Primitive("-"), "*"=.Primitive("*"))

ReadInputData <- function(conn) {
    line <- unlist(strsplit(readLines(conn, n=1), split = ""))
    len <- length(line)
    digits <- as.integer(line[seq(from = 1, to = len, by = 2)])
    operations <- line[seq(from = 2, to = len, by = 2)]
    list(digits=digits, operations=operations)
}

CalcMinMax <- function(ix, jy, operations, minimumOfSubexpressions, maximumOfSubexpressions) {
    subexpressionMin <- Inf
    subexpressionMax <- -Inf
    for (k in ix:(jy-1)) {
        op <- Operations[[operations[k]]]
        a <- op(maximumOfSubexpressions[ix, k], maximumOfSubexpressions[k+1, jy])
        b <- op(maximumOfSubexpressions[ix, k], minimumOfSubexpressions[k+1, jy])
        c <- op(minimumOfSubexpressions[ix, k], maximumOfSubexpressions[k+1, jy])
        d <- op(minimumOfSubexpressions[ix, k], minimumOfSubexpressions[k+1, jy])
        subexpressionMin <- min(subexpressionMin, a, b, c, d)
        subexpressionMax <- max(subexpressionMax, a, b, c, d)
    }
    list(maximum=subexpressionMax, minimum=subexpressionMin)
}

MaximizeArithmeticExpression <- function(digits, operations) {
    n <- length(digits)
    minimumOfSubexpressions <- matrix(data = 0, nrow = n, ncol = n)
    maximumOfSubexpressions <- matrix(data = 0, nrow = n, ncol = n)
    diag(minimumOfSubexpressions) <- digits
    diag(maximumOfSubexpressions) <- digits
    for (s in 1:(n-1)) {
        for (ix in 1:(n-s)) {
            jy <- ix + s
            extremeValues <- CalcMinMax(ix, jy, operations, minimumOfSubexpressions, maximumOfSubexpressions)
            minimumOfSubexpressions[ix, jy] <- extremeValues$minimum
            maximumOfSubexpressions[ix, jy] <- extremeValues$maximum
        }
    }
    maximumOfSubexpressions[1, n]
}

Main <- function() {
    conn <- file("stdin", "r")
    inputData <- ReadInputData(conn)
    close(conn)
    result <- MaximizeArithmeticExpression(inputData$digits, inputData$operations)
    cat(format(result, digits = 18), fill = TRUE)
}

if (!interactive()) {
    Main()
}

File: Haskell/algorithmic_toolbox/25-ArithmeticExpression.hs
import Data.List (partition)
import Data.Char (isDigit, digitToInt)
import Data.Array (Array(..), listArray, (!), range)
import Data.Map.Strict (Map(..), fromList)
import qualified Data.Map.Strict as Map

operationsDict :: Map Char (Int -> Int -> Int)
operationsDict = fromList [('+', (+)), ('*', (*)), ('-', (-))]


readInputData :: String -> (Array Int Int, Array Int Char)
readInputData line =
    let (digits, operations) = partition isDigit line
        nrDigits = length digits
        digitsArray = listArray (0, nrDigits - 1) (map digitToInt digits)
        operationsArray = listArray (0, nrDigits - 2) operations
    in (digitsArray, operationsArray)


calcMinMax :: Int -> Int -> Array Int Char -> Array (Int, Int) Int -> Array (Int, Int) Int -> (Int, Int)
calcMinMax ix jy operations minimumOfSubexpressions maximumOfSubexpressions =
    foldl calcSubExpressionValuesInRange (maxBound, minBound) [ix..(jy-1)]
    where
        calcSubExpressionValuesInRange :: (Int, Int) -> Int -> (Int, Int)
        calcSubExpressionValuesInRange acc k =
            let (subExpressionMin, subExpressionMax) = acc
                op = operationsDict Map.! (operations ! k)
                subexpressionsSplitAtK =
                    [op (maximumOfSubexpressions ! (ix, k)) (maximumOfSubexpressions ! (k + 1, jy)),
                     op (maximumOfSubexpressions ! (ix, k)) (minimumOfSubexpressions ! (k + 1, jy)),
                     op (minimumOfSubexpressions ! (ix, k)) (maximumOfSubexpressions ! (k + 1, jy)),
                     op (minimumOfSubexpressions ! (ix, k)) (minimumOfSubexpressions ! (k + 1, jy))]
                (splitMin, splitMax) = (minimum subexpressionsSplitAtK, maximum subexpressionsSplitAtK)
            in (min subExpressionMin splitMin, max subExpressionMax splitMax)


maximizeExpression :: Array Int Int -> Array Int Char -> Int
maximizeExpression digits operations = maximumOfSubexpressions ! (0, n - 1)
    where
        calcExtremeValues :: Int -> Int -> (Int, Int)
        calcExtremeValues ix jy
            | ix == jy = let digit = digits ! ix in (digit, digit)
            | otherwise = calcMinMax ix jy operations minimumOfSubexpressions  maximumOfSubexpressions
        n = length digits
        bounds = ((0, 0), (n - 1, n - 1))
        minimumOfSubexpressions = listArray bounds [fst $ calcExtremeValues i j | (i, j) <- range bounds]
        maximumOfSubexpressions = listArray bounds [snd $ calcExtremeValues i j | (i, j) <- range bounds]


main :: IO()
main = do
    line <- getLine
    let (digits, operations) = readInputData line
    print $ maximizeExpression digits operations
